File: Buddy/buddy-flutter/lib/core/services/achievement_unlocker.dart
import 'package:shared_preferences/shared_preferences.dart';

import '../../models/achievement.dart';
import '../../models/pet.dart';
import '../../models/pet_action.dart';
import '../../models/personality.dart';
import '../persistence/coin_wallet.dart';
import 'toast_queue.dart';

/// Centraliza la detección de achievements.
class AchievementUnlocker {
  static Future<void> checkAfterAction({
    required PetAction action,
    required Pet pet,
    required PersonalityTracker personality,
  }) async {
    final unlocked = <Achievement>[];

    Future<void> tryUnlock(Achievement a) async {
      if (await AchievementStore.unlock(a)) unlocked.add(a);
    }

    switch (action) {
      case PetAction.eat:
        await tryUnlock(Achievement.firstFeed);
        if (personality.feedCount >= 10) await tryUnlock(Achievement.fed10);
        if (personality.feedCount >= 50) await tryUnlock(Achievement.fed50);
        break;
      case PetAction.play:
        await tryUnlock(Achievement.firstPlay);
        if (personality.playCount >= 25) await tryUnlock(Achievement.played25);
        break;
      case PetAction.sleep:
        await tryUnlock(Achievement.firstSleep);
        break;
      default:
        break;
    }

    if (pet.moodLevel >= 3) await tryUnlock(Achievement.maxMood);

    final s = pet.stats;
    if (s.hunger > 90 && s.thirst > 90 && s.energy > 90 && s.hygiene > 90 && s.happiness > 90) {
      await tryUnlock(Achievement.fullStats);
    }

    final coins = await CoinWallet.balance();
    if (coins >= 100) await tryUnlock(Achievement.coins100);
    if (coins >= 500) await tryUnlock(Achievement.coins500);

    if (pet.ageInDays >= 3)  await tryUnlock(Achievement.daysAlive3);
    if (pet.ageInDays >= 7)  await tryUnlock(Achievement.daysAlive7);
    if (pet.ageInDays >= 30) await tryUnlock(Achievement.daysAlive30);

    for (final a in unlocked) {
      ToastQueue.instance.show(emoji: a.emoji, title: 'Logro: ${a.title}', detail: '+${a.reward}🪙');
    }
  }

  static Future<void> bath() async {
    if (await AchievementStore.unlock(Achievement.firstBath)) {
      ToastQueue.instance.show(
        emoji: Achievement.firstBath.emoji,
        title: 'Logro: ${Achievement.firstBath.title}',
        detail: '+${Achievement.firstBath.reward}🪙',
      );
    }
  }

  static Future<void> minigamePlayed() async {
    if (await AchievementStore.unlock(Achievement.minigame1)) {
      ToastQueue.instance.show(
        emoji: Achievement.minigame1.emoji,
        title: 'Logro: ${Achievement.minigame1.title}',
        detail: '+${Achievement.minigame1.reward}🪙',
      );
    }
  }

  static Future<void> reincarnation() async {
    if (await AchievementStore.unlock(Achievement.firstReincarnation)) {
      ToastQueue.instance.show(
        emoji: Achievement.firstReincarnation.emoji,
        title: 'Logro: ${Achievement.firstReincarnation.title}',
        detail: '+${Achievement.firstReincarnation.reward}🪙',
      );
    }
  }

  static Future<void> recordMinigamePlayed(String minigameId) async {
    final prefs = await SharedPreferences.getInstance();
    final played = (prefs.getStringList('buddy.minigames.played.v1') ?? const []).toSet()
      ..add(minigameId);
    await prefs.setStringList('buddy.minigames.played.v1', played.toList());
  }
}

File: Buddy/buddy-ios/Sources/buddy/Core/Services/AchievementUnlocker.swift
import Foundation

/// Centralizes achievement detection. Call after every relevant event.
struct AchievementUnlocker {
    static func checkAfterAction(_ action: PetAction, pet: Pet, personality: PersonalityTracker) {
        var unlocked: [Achievement] = []
        // First-time actions
        switch action {
        case .eat:
            if AchievementStore.unlock(.firstFeed) { unlocked.append(.firstFeed) }
            if personality.feedCount >= 10, AchievementStore.unlock(.fed10) { unlocked.append(.fed10) }
            if personality.feedCount >= 50, AchievementStore.unlock(.fed50) { unlocked.append(.fed50) }
        case .play:
            if AchievementStore.unlock(.firstPlay) { unlocked.append(.firstPlay) }
            if personality.playCount >= 25, AchievementStore.unlock(.played25) { unlocked.append(.played25) }
        case .sleep:
            if AchievementStore.unlock(.firstSleep) { unlocked.append(.firstSleep) }
        default: break
        }
        // Stats milestones
        if pet.moodLevel >= 3 {
            if AchievementStore.unlock(.maxMood) { unlocked.append(.maxMood) }
        }
        let s = pet.stats
        if s.hunger > 90 && s.thirst > 90 && s.energy > 90 && s.hygiene > 90 && s.happiness > 90 {
            if AchievementStore.unlock(.fullStats) { unlocked.append(.fullStats) }
        }
        // Coin milestones
        if CoinWallet.balance >= 100, AchievementStore.unlock(.coins100) { unlocked.append(.coins100) }
        if CoinWallet.balance >= 500, AchievementStore.unlock(.coins500) { unlocked.append(.coins500) }
        // Age milestones
        switch pet.ageInDays {
        case 3...:  if AchievementStore.unlock(.daysAlive3) { unlocked.append(.daysAlive3) }
        default: break
        }
        if pet.ageInDays >= 7,  AchievementStore.unlock(.daysAlive7)  { unlocked.append(.daysAlive7)  }
        if pet.ageInDays >= 30, AchievementStore.unlock(.daysAlive30) { unlocked.append(.daysAlive30) }

        for ach in unlocked {
            ToastQueue.shared.show(emoji: ach.emoji, title: "Logro: \(ach.title)", detail: "+\(ach.reward)🪙")
        }
    }

    static func bath() {
        if AchievementStore.unlock(.firstBath) {
            ToastQueue.shared.show(emoji: Achievement.firstBath.emoji, title: "Logro: \(Achievement.firstBath.title)", detail: "+\(Achievement.firstBath.reward)🪙")
        }
    }

    static func minigamePlayed() {
        if AchievementStore.unlock(.minigame1) {
            ToastQueue.shared.show(emoji: Achievement.minigame1.emoji, title: "Logro: \(Achievement.minigame1.title)", detail: "+\(Achievement.minigame1.reward)🪙")
        }
        let played = UserDefaults.standard.stringArray(forKey: "buddy.minigames.played.v1") ?? []
        if Set(played).count >= MinigameID.allCases.count, AchievementStore.unlock(.allMinigames) {
            ToastQueue.shared.show(emoji: Achievement.allMinigames.emoji, title: "Logro: \(Achievement.allMinigames.title)", detail: "+\(Achievement.allMinigames.reward)🪙")
        }
    }

    static func reincarnation() {
        if AchievementStore.unlock(.firstReincarnation) {
            ToastQueue.shared.show(emoji: Achievement.firstReincarnation.emoji, title: "Logro: \(Achievement.firstReincarnation.title)", detail: "+\(Achievement.firstReincarnation.reward)🪙")
        }
    }

    static func recordMinigamePlayed(_ id: MinigameID) {
        var played = Set(UserDefaults.standard.stringArray(forKey: "buddy.minigames.played.v1") ?? [])
        played.insert(id.rawValue)
        UserDefaults.standard.set(Array(played), forKey: "buddy.minigames.played.v1")
    }
}
